radix-engine/src/blueprints/intent_hash_store/package.rs

```diff
@@ -0,0 +1,280 @@
+        let max_epoch_exclusive =
+            self.start_epoch + num_partitions as u64 * self.epochs_per_partition;
+        if epoch < self.start_epoch || epoch >= max_epoch_exclusive {
+            return None;
+        }
+
+        // Calculate the destination partition number
+        let mut partition_number =
+            self.start_partition as u64 + (epoch - self.start_epoch) / self.epochs_per_partition;
+        if partition_number > self.partition_range_end as u64 {
+            partition_number -= num_partitions as u64;
+        }
+
+        assert!(partition_number >= self.partition_range_start as u64);
+        assert!(partition_number <= self.partition_range_end as u64);
+
+        Some(partition_number as u8)
+    }
+
+    pub fn advance(&mut self) -> u8 {
+        let old_start_partition = self.start_partition;
+        self.start_epoch += self.epochs_per_partition;
+        self.start_partition = if self.start_partition == self.partition_range_end {
+            self.partition_range_start
+        } else {
+            self.start_partition + 1
+        };
+        old_start_partition
+    }
+}
+
+pub struct IntentHashStoreBlueprint;
+
+impl IntentHashStoreBlueprint {
+    pub fn create<Y>(
+        address_reservation: GlobalAddressReservation,
+        api: &mut Y,
+    ) -> Result<GlobalAddress, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        let current_epoch = Runtime::current_epoch(api)?;
+        let intent_store = api.new_simple_object(
+            INTENT_HASH_STORE_BLUEPRINT,
+            vec![scrypto_encode(&IntentHashStoreSubstate {
+                start_epoch: current_epoch.number(),
+                start_partition: PARTITION_RANGE_START,
+                partition_range_start: PARTITION_RANGE_START,
+                partition_range_end: PARTITION_RANGE_END,
+                epochs_per_partition: EPOCHS_PER_PARTITION,
+            })
+            .unwrap()],
+        )?;
+        let access_rules = AccessRules::create(Roles::new(), api)?.0;
+        let metadata = Metadata::create(api)?;
+        let royalty = ComponentRoyalty::create(RoyaltyConfig::default(), api)?;
+
+        let address = api.globalize_with_address(
+            btreemap!(
+                ObjectModuleId::Main => intent_store,
+                ObjectModuleId::AccessRules => access_rules.0,
+                ObjectModuleId::Metadata => metadata.0,
+                ObjectModuleId::Royalty => royalty.0,
+            ),
+            address_reservation,
+        )?;
+        Ok(address)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn calculate_coverage() {
+        let covered_days = EPOCHS_PER_PARTITION as f64
+            * (PARTITION_RANGE_END as f64 - (PARTITION_RANGE_START as f64 - 1.0) - 1.0)
+            * 5.0
+            / 60.0
+            / 24.0;
+
+        assert_eq!(covered_days.floor() as usize, 65);
```

Comment:
Can we add an assertion that this is greater than the default epoch length from the transaction validator?

Reply:
Yes, will add!